### DISCUSSION TITLE
when there are multiple rust binaries

### DISCUSSION THREAD

**Discussion by laoshaw** [johnthagen/min-sized-rust#75] [General]
they will also carry the same stdlib and adds up the size fast, I wonder why can't rust have a shared library of stdlib just like libstdc++? to win from c++, you can't have this large size when you have 100+ executables in one system, it will blow up the storage fast on storage restricted systems.

**Comment by NobodyXu** [comment]
The problem is that rust ABI is unstable and change for optimization reasons.

I recommend using multi-call binary, like how busybox does, you would not only get shared stdlib but also dependencies duplication and other code deduplication.

**Comment by laoshaw** [comment]
if I rebuild my rust code against the same rust stdlib(libstd...so), and when the libstd changes I will recompile all my code with -C prefer-dynamic, that will work just like c++? that is, one shared lib to be used by multiple rust programs?

**Reply by NobodyXu** [reply]
That will only work if they are compiled by the same version of rustc/cargo, it works but is not guaranteed.

Rust can technically change ABI in every compilation, there's a nightly flag to change ABI every compilation for testing purpose.

**Reply by laoshaw** [reply]
if I use the same rustc/cargo comes along with the stable released stdlib to build all my own programs, they're supposed to always work? what do you mean it's not 'guaranteed', is it because 'prefer-dynamic' not well tested by rust team and thus not 'production ready'?

**Reply by NobodyXu** [reply]
> what do you mean it's not 'guaranteed'

It's rustc/cargo does not provide any guarantee.

Yes, if you use the stable rustc/cargo now it works, but they can change it at anytime and it's not considered to be a breaking change.

**Reply by NobodyXu** [reply]
Given that you might also have duplicate dependencies, I recommend to create multi-call binary which would be much smaller.
